### PR TITLE
Make validator stable when beacon node goes offline

### LINKED ISSUE #8188: Validator Client crash-loops until beacon is accessible and sync'd (to beacon chain?)
# 🚀 Feature Request

### Description

When we deploy a new eth2 cluster (redundant beacons and multiple clients) to our platform, the beacon node takes some time for its DNS name to resolve; until this resolves, the client crash-loops failing to connect.
Once it can connect, the client crash-loops for several minutes more until the beacon node is sync'd (presumably to the eth2 beacon chain).  This crash-looping may be the intended behavior but it's not user-friendly or platform-friendly.

e.g. after connection made:
```
time="2021-01-02 21:05:34" level=info msg="Waiting for beacon chain start log from the ETH 1.0 deposit contract" prefix=validator
time="2021-01-02 21:10:36" level=fatal msg="Could not determine if beacon chain started: could not receive ChainStart from stream: rpc error: code = Unavailable desc = transport is closing" prefix=validator
```

When it finally stops crashing:
```
time="2021-01-02 21:15:36" level=info msg="Waiting for beacon chain start log from the ETH 1.0 deposit contract" prefix=validator
time="2021-01-02 21:15:36" level=info msg="Beacon chain started" genesisTime=2020-12-01 12:00:23 +0000 UTC prefix=validator
```

### Describe the solution you'd like

A client should be able to be started before its beacon node and be resilient and patient.  If nothing else, this behavior should be optional (`--no-fail-fast`).

### Describe alternatives you've considered

If we can't get this added, we'll need to build our own readiness probes to hold off starting the client until it's likely not going to crash.  Pod crash-loops are something our platforms alerts on as, in general, crashing under 'normal operation' should be avoided.

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
Currently validator client crashes before entering the mail loop. it can crash on few cases:
- `WaitForChainStart` connection issue
- `WaitForSync` connection issue
- 'WaitForActivation'  has inner retry flow as well (should i remove it?)
- `CanonicalHeadSlot` connection issue

while investigating the right way to implement the fail safe mechanism for beacon client availability i found a case where validator could stop getting blocks
I have marked this problematic line with a comment

I tested it in runtime while crushing the beacon node on several cases (looks stable)

This pr polls the endpoint of beacon node while it is nor reachable and waits till its up to get on with its functionality

**Which issues(s) does this PR fix?**

Fixes #8188

### CODE REVIEW

File: validator/client/runner.go [hunk 1]
```diff
@@ -62,25 +67,52 @@ func run(ctx context.Context, v Validator) {
...
+				continue
+			}
+			if err != nil {
+				log.Fatalf("Could not get current canonical head slot: %v", err)
+			}
+		}
+		break
 	}
+	connectionErrorChannel := make(chan error)
+	go v.ReceiveBlocks(ctx, connectionErrorChannel)
```
Comment (shayzluf) [inline validator/client/runner.go:112]: there was no error handling for ReceiveBlocks while it was running in a separate goroutine

File: validator/client/runner.go [hunk 2]
```diff
@@ -62,25 +67,52 @@ func run(ctx context.Context, v Validator) {
...
+				log.Warnf("Could not get current canonical head slot: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Fatalf("Could not get current canonical head slot: %v", err)
+			}
+		}
+		break
 	}
+	connectionErrorChannel := make(chan error)
```
Comment (rauljordan) [inline validator/client/runner.go:111]: I think this should be a buffered chan (with 1 as the buffer)

File: validator/client/runner.go [hunk 3]
```diff
@@ -62,25 +67,52 @@ func run(ctx context.Context, v Validator) {
...
+			err = v.WaitForActivation(ctx)
+			if err != nil && strings.Contains(err.Error(), connectionIssueMsg) {
+				log.Warnf("Could not wait for validator activation: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Fatalf("Could not wait for validator activation: %v", err)
+			}
+			headSlot, err = v.CanonicalHeadSlot(ctx)
+			if err != nil && strings.Contains(err.Error(), connectionIssueMsg) {
```
Comment (rauljordan) [inline validator/client/runner.go:101]: if we could use errors.Is that would be cleaner

File: validator/client/runner.go [hunk 4]
```diff
@@ -62,25 +67,52 @@ func run(ctx context.Context, v Validator) {
...
+	for {
+		select {
+		case <-ticker.C:
+			err := v.WaitForChainStart(ctx)
+			if err != nil && strings.Contains(err.Error(), connectionIssueMsg) {
+				log.Warnf("Could not determine if beacon chain started: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Fatalf("Could not determine if beacon chain started: %v", err)
```
Comment (rauljordan) [inline validator/client/runner.go:82]: Do we ned to fatal and can we instead do the backoff continue?

File: validator/client/runner.go [hunk 5]
```diff
@@ -62,25 +67,52 @@ func run(ctx context.Context, v Validator) {
...
+			}
+			err = v.WaitForSync(ctx)
+			if err != nil && strings.Contains(err.Error(), connectionIssueMsg) {
+				log.Warnf("Could not determine if beacon chain started: %v", err)
+				continue
+			}
+			if err != nil {
+				log.Fatalf("Could not determine if beacon node synced: %v", err)
```
Comment (rauljordan) [inline validator/client/runner.go:90]: Same here

File: validator/client/runner.go [hunk 6]
```diff
@@ -15,6 +16,10 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+// time to wait before trying to reconnect with the eth1 node.
+var backOffPeriod = 15 * time.Second
+var connectionIssueMsg = ", beacon chain may be offline..."
```
Comment (rauljordan) [inline validator/client/runner.go:21]: This message string is a bit awkward starting with the ",". Also if it is an error, we should name it errConnectionIssue

File: validator/client/runner.go [hunk 7]
```diff
@@ -15,6 +16,10 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+// time to wait before trying to reconnect with the eth1 node.
+var backOffPeriod = 15 * time.Second
```
Comment (rauljordan) [inline validator/client/runner.go:20]: 10 seconds could be better here, no harm in reducing it a bit more

File: validator/client/runner.go (hunk 6 above)
Comment (nisdas) [reply validator/client/runner.go:21]: +1 there is a common pattern to naming errors:
```
errConnectionIssue
```
etc

File: validator/client/runner.go [hunk 8]
```diff
@@ -89,6 +121,12 @@ func run(ctx context.Context, v Validator) {
 			log.Info("Context canceled, stopping validator")
 			span.End()
 			return // Exit if context is canceled.
+		case blocksError := <-connectionErrorChannel:
+			if blocksError != nil && strings.Contains(blocksError.Error(), connectionIssueMsg) {
+				go v.ReceiveBlocks(ctx, connectionErrorChannel)
+				continue
+			}
+			log.WithError(blocksError).Fatal("block stream interrupted")
```
Comment (nisdas) [inline validator/client/runner.go:129]: should this be fatal ? @terencechain you introduced the block receiver in #8197, should the error from there be fatal or
can we continue running the validator client fine.

File: validator/client/validator.go [hunk 9]
```diff
@@ -133,7 +133,7 @@ func (v *validator) WaitForChainStart(ctx context.Context) error {
 	// First, check if the beacon chain has started.
 	stream, err := v.validatorClient.WaitForChainStart(ctx, &ptypes.Empty{})
 	if err != nil {
-		return errors.Wrap(err, "could not setup beacon chain ChainStart streaming client")
+		return errors.Wrap(err, "could not setup beacon chain ChainStart streaming client"+connectionIssueMsg)
```
Comment (nisdas) [inline validator/client/validator.go:136]: maybe there could be a cleaner way to wrap the connection error here. You could have a small function which would wrap the underlying error with the connection error for all the cases.

File: validator/client/validator.go [hunk 10]
```diff
@@ -241,10 +243,11 @@ func (v *validator) SlasherReady(ctx context.Context) error {
 // ReceiveBlocks starts a gRPC client stream listener to obtain
 // blocks from the beacon node. Upon receiving a block, the service
 // broadcasts it to a feed for other usages to subscribe to.
-func (v *validator) ReceiveBlocks(ctx context.Context) {
+func (v *validator) ReceiveBlocks(ctx context.Context, connectionErrorChannel chan error) {
 	stream, err := v.beaconClient.StreamBlocks(ctx, &ethpb.StreamBlocksRequest{VerifiedOnly: true})
 	if err != nil {
-		log.WithError(err).Error("Failed to retrieve blocks stream")
+		log.WithError(err).Error("Failed to retrieve blocks stream" + connectionIssueMsg)
+		connectionErrorChannel <- err
```
Comment (nisdas) [inline validator/client/validator.go:250]: still not sure if a channel is required over here, will let terence answer on how best to handle failures over here

File: validator/client/validator.go (hunk 9 above)
Comment (shayzluf) [reply validator/client/validator.go]: fixed please take another look

File: validator/client/runner.go [hunk 11]
```diff
@@ -62,33 +67,77 @@ func run(ctx context.Context, v Validator) {
...
-
 	for {
+		slotCtx, cancel := context.WithCancel(ctx)
 		ctx, span := trace.StartSpan(ctx, "validator.processSlot")
 
 		select {
 		case <-ctx.Done():
 			log.Info("Context canceled, stopping validator")
 			span.End()
+			cancel()
```
Comment (shayzluf) [inline validator/client/runner.go:133]: missing call to cancel? @rauljordan goland detected a possible context leak

File: validator/client/runner.go (hunk 8 above)
Comment (rauljordan) [reply validator/client/runner.go:145]: cc @terencechain

**Review by rauljordan — APPROVED** [review]
This all makes sense to me, nice tests @shayzluf - I think this will be a very nice improvement

File: validator/client/runner.go (hunk 8 above)
Comment (terencechain) [reply validator/client/runner.go:145]: Continue running is fine. The worst case is it'll wait 4s to attest

File: validator/client/runner.go [hunk 12]
```diff
@@ -64,37 +69,80 @@ func run(ctx context.Context, v Validator) {
...
+		headSlot, err = v.CanonicalHeadSlot(ctx)
+		if err != nil && errors.Is(err, errConnectionIssue) {
+			log.Warnf("Could not get current canonical head slot: %v", err)
+			continue
+		}
+		if err != nil {
+			log.Fatalf("Could not get current canonical head slot: %v", err)
+		}
+		done = true
+		select {
```
Comment (terencechain) [inline validator/client/runner.go:111]: Do we really need this select? Can't we just add `ctx.Err()` check at the top?
```
if ctx.Err() != nil {
		return nil, ctx.Err()
	}
``` 
and break once we get to the end of the loop

I think that is simpler to reason?

File: validator/client/runner.go [hunk 13]
```diff
@@ -16,6 +17,10 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+// time to wait before trying to reconnect with the eth1 node.
+var backOffPeriod = 10 * time.Second
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply validator/client/runner.go:21]: the issue is in the comment this is a general connection issue back off (not eth1)
thanks for pointing out

File: validator/client/runner.go (hunk 12 above)
Comment (shayzluf) [reply validator/client/runner.go]: tried to make it clearer

**Review by terencechain — APPROVED** [review]

**Pull request merged**